test/unit-tests/sourcekit-lsp/LanguageClientManager.test.ts

```diff
@@ -0,0 +1,408 @@
+//
+// This source file is part of the VS Code Swift open source project
+//
+// Copyright (c) 2024 the VS Code Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VS Code Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as assert from "assert";
+import * as vscode from "vscode";
+import { FolderEvent, WorkspaceContext } from "../../../src/WorkspaceContext";
+import {
+    LanguageClientError,
+    LanguageClientManager,
+    SourceKitLSPErrorHandler,
+} from "../../../src/sourcekit-lsp/LanguageClientManager";
+import {
+    anyFunction,
+    anything,
+    instance,
+    mock,
+    when,
+    verify,
+    anyString,
+    capture,
+} from "ts-mockito";
+import { Version } from "../../../src/utilities/version";
+import { SwiftToolchain } from "../../../src/toolchain/toolchain";
+import { BuildFlags } from "../../../src/toolchain/BuildFlags";
+import configuration from "../../../src/configuration";
+import { SwiftOutputChannel } from "../../../src/ui/SwiftOutputChannel";
+import { eventListenerMock, mockValue, mockConstructor, mockNamespace } from "../MockUtils";
+import * as languageClient from "vscode-languageclient/node";
+import { FolderContext } from "../../../src/FolderContext";
+
+suite("LanguageClientManager Suite", () => {
+    let workspace: WorkspaceContext;
+    let toolchain: SwiftToolchain;
+    let sut: LanguageClientManager;
+
+    const languageClientMock = mockConstructor(languageClient, "LanguageClient");
+    const onDidChangeStateMock = eventListenerMock(languageClientMock, "onDidChangeState");
+    const lspConfig = mockValue(configuration, "lsp");
+
+    setup(async () => {
+        workspace = mock(WorkspaceContext);
+        toolchain = mock(SwiftToolchain);
+        when(workspace.swiftVersion).thenReturn(new Version(6, 0, 0));
+        when(toolchain.getToolchainExecutable("sourcekit-lsp")).thenReturn(
+            "/path/to/toolchain/bin/sourcekit-lsp"
+        );
+        const toolchainInstance = instance(toolchain);
+        when(toolchain.buildFlags).thenReturn(new BuildFlags(toolchainInstance));
+        when(workspace.toolchain).thenReturn(toolchainInstance);
+        when(workspace.outputChannel).thenReturn(new SwiftOutputChannel("Swift Test", false, 100));
+        when(workspace.subscriptions).thenReturn([]);
+        when(workspace.observeFolders(anything())).thenReturn(new vscode.Disposable(() => {}));
+
+        when(languageClientMock.onNotification(anything(), anyFunction())).thenReturn(
+            new vscode.Disposable(() => {})
+        );
+        when(languageClientMock.clientOptions).thenReturn({});
+        when(languageClientMock.start).thenReturn(() => Promise.resolve());
+    });
+
+    suite("state", () => {
+        setup(() => {});
```

Comment:
nit: required?

Reply:
guess not, but going to leave since it's there